source/sigv4.c

```diff
@@ -2913,38 +2978,37 @@ SigV4Status_t SigV4_GenerateHTTPAuthorization( const SigV4Parameters_t * pParams
     char * pSignedHeaders = NULL;
     size_t encodedLen = 0U, algorithmLen = 0U, signedHeadersLen = 0U, authPrefixLen = 0U;
     HmacContext_t hmacContext = { 0 };
+
     SigV4String_t signingKey;
     ptrdiff_t bufferLen;
 
-    returnStatus = verifySigV4Parameters( pParams );
-
-    authPrefixLen = *authBufLen;
-
-    /* Default arguments. */
-    if( ( pParams->pAlgorithm == NULL ) || ( pParams->algorithmLen == 0U ) )
-    {
-        /* The default algorithm is AWS4-HMAC-SHA256. */
-        pAlgorithm = SIGV4_AWS4_HMAC_SHA256;
-        algorithmLen = SIGV4_AWS4_HMAC_SHA256_LENGTH;
-    }
-    else
-    {
-        pAlgorithm = pParams->pAlgorithm;
-        algorithmLen = pParams->algorithmLen;
-    }
+    returnStatus = verifyParamsToGenerateAuthHeaderApi( pParams,
+                                                        pAuthBuf, authBufLen,
+                                                        pSignature, signatureLen );
 
     if( returnStatus == SigV4Success )
     {
-        returnStatus = generateCanonicalRequestUntilHeaders( pParams, &canonicalContext, &pSignedHeaders, &signedHeadersLen );
+        authPrefixLen = *authBufLen;
+
+        /* Default arguments. */
+        if( ( pParams->pAlgorithm == NULL ) || ( pParams->algorithmLen == 0U ) )
+        {
+            /* The default algorithm is AWS4-HMAC-SHA256. */
+            pAlgorithm = SIGV4_AWS4_HMAC_SHA256;
+            algorithmLen = SIGV4_AWS4_HMAC_SHA256_LENGTH;
+        }
+        else
+        {
+            pAlgorithm = pParams->pAlgorithm;
+            algorithmLen = pParams->algorithmLen;
+        }
```

Comment:
If the complexity of this function is too high, you can reduce it by moving this code out of the `if( returnStatus == SigV4Success )` block

Reply:
Because I did this already:
```
/**
 * @brief Verify all authorization-related parameters passed to
 * #SigV4_GenerateHTTPAuthorization.
 *
 * @param[in] pParams Complete SigV4 configurations passed by application.
 * @param[out] pAlgorithm The algorithm used for SigV4 authentication.
 * @param[out] algorithmLen The length of @p pAlgorithm.
 */
static void assignDefaultArguments( const SigV4Parameters_t * pParams,
                                    char ** pAlgorithm,
                                    size_t * algorithmLen );
```

```
static void assignDefaultArguments( const SigV4Parameters_t * pParams,
                                    char ** pAlgorithm,
                                    size_t * algorithmLen )
{
    if( ( pParams->pAlgorithm == NULL ) || ( pParams->algorithmLen == 0 ) )
    {
        /* The default algorithm is AWS4-HMAC-SHA256. */
        *pAlgorithm = SIGV4_AWS4_HMAC_SHA256;
        *algorithmLen = SIGV4_AWS4_HMAC_SHA256_LENGTH;
    }
    else
    {
        *pAlgorithm = pParams->pAlgorithm;
        *algorithmLen = pParams->algorithmLen;
    }
}
```